packages/core/src/internal-plugin/theme.ts

```diff
@@ -18,16 +18,46 @@ export const themeToolCtx = createSlice<ThemeTool>(
 export const emotionConfigCtx = createSlice<Options>({ key: 'milkdown' }, 'EmotionConfig');
 export const emotionCtx = createSlice<Emotion>({} as Emotion, 'Emotion');
 
+export type ThemeProvider<T = undefined> = (themeTool: ThemeTool, info?: T) => string | void;
+export type ThemeProviderKey<T = undefined> = Slice<ThemeProvider<T>>;
+
+export const createThemeProviderKey = <T = undefined>(key = 'themeComponentKey'): ThemeProviderKey<T> =>
+    createSlice((() => '') as ThemeProvider<T>, key);
+
+export type ThemeManager = {
+    provide: <T>(meta: ThemeProviderKey<T>, value: ThemeProvider<T>) => void;
+    consume: <T>(meta: ThemeProviderKey<T> | string, info?: T) => ReturnType<ThemeProvider<T>>;
+};
+
+export const themeManagerCtx = createSlice<ThemeManager>({} as ThemeManager, 'themeManager');
+
 export type { Emotion, ThemeTool } from '@milkdown/design-system';
 
+export const themeTimerCtx = createSlice<Timer[]>([], 'themeTimer');
+export const ThemeReady = createTimer('ThemeReady');
```

Comment:
I've seen the time and `xReady` pattern in the codebase quite a bit. Is this to manage the case where plugins might be asynchronous and need to report back when they're done loading? Why'd you go this route instead of passing promises?

Reply:
With this pattern you can write plugin to `defer` other plugin's loading status.

For example, let's say you have plugin A and B. B will wait for A loaded. And now we want to create a custom plugin C, which will let B not only wait for A, but also wait for C.

Thus, it will make following changes:

BTimerCtx: [AReady]  -----> [AReady, CReady].

That's how it works. Plugins may expose a timer that let other plugins control it's ready status.